### ISSUE TITLE
memory leak in relay and aggregator daemons

### ISSUE THREAD

**Issue by dkulikovsky** [graphite-project/carbon#185]
I've got a leaking megacarbon instances. The load is low - 50-100k metrics per minute. Config is simple - i have a pipeline 2 relays with aggregated-consistent-hashing -> 1 aggregator with consistent hashing -> 2 writers. 
Instances that run relay or aggregation role consumed tons of memory:

```
top - 16:03:22 up 6 days,  1:09,  7 users,  load average: 143.34, 76.86, 48.83
Tasks: 414 total,   4 running, 410 sleeping,   0 stopped,   0 zombie
Cpu(s):  8.2%us,  0.5%sy,  0.0%ni, 90.0%id,  1.2%wa,  0.0%hi,  0.1%si,  0.0%st
Mem:  65949272k total, 65443728k used,   505544k free,  1215012k buffers
Swap:        0k total,        0k used,        0k free,   269768k cached

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  WCHAN     COMMAND                                                                                                                        
 2824 www-data  20   0 23.7g  23g 1816 R 99.4 37.5  34:38.02 -         /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-relay-st01 start                                                  
 2897 www-data  20   0 11.9g  11g 1760 R 99.4 18.8   9:28.29 -         /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-relay-st02 start                                                  
15538 www-data  20   0 23.7g  23g 1796 S 44.7 37.5  14:47.56 ep_poll   /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-aggregator-st01 start                                             
 1125 www-data  20   0  299m 121m 1860 S  7.9  0.2  31:58.85 poll_sche /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-writer-st02 start                                                 
 1120 www-data  20   0  310m 146m 1860 S  7.4  0.2  37:09.29 poll_sche /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-writer-st01 start                                                 
 1118 www-data  20   0  310m 146m 1860 S  3.2  0.2  20:11.00 ep_poll   /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-writer-st01 start                                                 
 1124 www-data  20   0  299m 121m 1860 S  3.2  0.2  17:23.44 ep_poll   /usr/bin/python /var/lib/graphite/bin/carbon-daemon.py carbon-writer-st02 start               
```

That happens really fast, just after restarting megacarbon daemons and starting perf test tool, in first minute relay and aggregator consume about 720Mb of memory each, in the second minute memory usage grows to 3Gb each. After 20-30 min procs consume 6-11Gb each, and so on. Top output above is taken after 10hours after restarting carbon daemons (perf test lasted for 30min and then all these daemons works only on data it generates by itself).
If RELAY_METHOD is set to relay-rules for relay instances - everything is fine, memory usage is about 30-50Mb, but for aggregator this doesn't help at all.

I've tried to catch this memory leak with different tools, but i didn't succeed.

Current config for relay is
```
REPLICATION_FACTOR = 1
MAX_DATAPOINTS_PER_MESSAGE = 5000
MAX_QUEUE_SIZE = 1000
USE_FLOW_CONTROL = True
DESTINATIONS = 127.0.0.1:12301:st01
RELAY_METHOD = aggregated-consistent-hashing
```
for aggregator:
```
REPLICATION_FACTOR = 1
MAX_DATAPOINTS_PER_MESSAGE = 5000
MAX_QUEUE_SIZE = 1000000
USE_FLOW_CONTROL = True
DESTINATIONS = 127.0.0.1:2301:st01, 127.0.0.1:2302:st02
RELAY_METHOD = consistent-hashing
```

**Comment by tristanbes** [comment]
I can confirm that too.

Server is crashing because 100% of memory is used by python.
What did you do to fix this please?

![2015-03-30_1136](https://cloud.githubusercontent.com/assets/346010/6893930/376b8cd8-d6d1-11e4-8f62-29e321b4c8a7.png)

**Comment by obfuscurity** [comment]
@jamesjuran @bruce-lyft Anyone have a patch suitable for a PR?

**Comment by mleinart** [comment]
Has be2e4f0d919b4cbbaf6743eebde4e41e3c775e73 improved things for anyone?

I'm not sure a bounded LRU cache makes a lot of sense in this case. Say we have a relay with this problem:
* 1 million metrics per minute are passing through
* Each of the metrics are sent once per minute (i.e. the relay sees each metric name once a minute)
* The metrics come in roughly the same order each time (each sender is sending 60sec from the last send and the senders are at a constant time difference from each other)

If we have an LRU cache that can hold 100,000 metrics, we cache the first 100k results, the next 100k boots those last 100k from the cache, the next 100k boots the prior... Basically we just eat memory and CPU and never get the benefit of the cache.

I can see some scenarios where perhaps a bounded cache might be some benefit, e.g. there is a wide distribution of metric frequencies sent through (some each at 1s, 10s, 60s, 10m, whatever) but that doesnt seem like the typical cases carbon relay/aggregator sees.

I also wonder how much help this cache actually is - perhaps someone with a rather large aggregator rule set could try removing it to see what happens to CPU usage?

**Comment by frlinux** [comment]
I never responded to that thread, apologies. The issue we had on our setup was with relays using hashing. And it was a bug in the python packages we used, which had a memory leak on the hashlib (https://lists.fedoraproject.org/pipermail/scm-commits/2012-July/832476.html)

After patching this, we never got a problem on relays, rock stable since as long as you do not feed them too many metrics.

Never used carbon aggregator so cannot comment on that one.

**Comment by bruce-lyft** [comment]
@mleinart The problem with the fix in be2e4f0 is that memory use is still unbounded. In our case, we have new metrics coming in on a regular basis, with a short time horizon (collectd/statsd metrics that are per-host, for ephemeral hosts, but which are aggregated; for example, CPU average for a group of hosts). With only your change we see carbon-aggregator memory use grow without bound and eventually crash.   I  believe that a change like the PR I submitted (https://github.com/graphite-project/carbon/pull/335) is still required.

**Comment by obfuscurity** [comment]
@mleinart @bruce-lyft Would it make sense to offer something like #335 as an optional strategy?

**Comment by obfuscurity** [comment]
I believe the aggregator leak in #185 is almost certainly the result of feeding aggregated metrics back into the aggregator loop as described in #455 and #560. There's a proposed fix in #439 but it needs some TLC before we can merge it. There's also an external fix applied to https://github.com/criteo-forks/carbon/commit/825693c97e57c7cdc36d0b0d758694010ce2be53 but I'd like to get feedback from others before we consider either of these.

Note that I'm not seeing any memory leaks in relays running at moderate volume (100k points/sec).

/cc @mleinart @drawks @iain-buclaw-sociomantic

**Comment by obfuscurity** [comment]
Adding to my previous comment, I believe that any previous relay memory leaks have already been fixed in master. The aggregator "leak" that @bruce-lyft describes is a consequence of the buffer dictionary memory footprint as it grows to accommodate the maximum concurrent number of buffers. Any existing memory should be reused, and my tests corroborate this hypothesis.

I agree with @mleinart's sentiment about the LRU cache not being the right approach. This would short-circuit Carbon's intended design and effectively trade memory for CPU consumption.

I have some thoughts about expiring metric buffers but I'm not sure how to tackle that yet.

Unless anyone has a reproducible case for memory leaks that aren't a natural byproduct of datapoints being stored in Python dictionaries, I propose that we close this issue as resolved. I acknowledge that there are times where Carbon memory could grow unbounded (e.g. `MAX_CACHE_SIZE = inf`) but these are tunable options (e.g. `MAX_CACHE_SIZE`, `MAX_QUEUE_SIZE`) and Carbon provides us with the internal statistics needed to identify these conditions before they become fatal.
